File: leetcode-js/255-verify-preorder-sequence-in-binary-search-tree.js
/**
 * @param {number[]} preorder
 * @return {boolean}
 */
var verifyPreorder = function(preorder) {
    'use strict';

    let low = Number.MIN_VALUE;
    const stack = [];

    for (const node of preorder) {
        if (node < low) {
            return false;
        }

        while (stack.length && node > stack[stack.length - 1]) {
            low = stack.pop();
        }

        stack.push(node);
    }

    return true;
};

File: leetcode-java/255-verify-preorder-sequence-in-binary-search-tree.java
public class Solution {
    public boolean verifyPreorder(int[] preorder) {
        int low = Integer.MIN_VALUE;
        Stack<Integer> stack = new Stack<Integer>();
        for (int val : preorder) {
            if (val < low) {
                return false;
            } else {
                while (!stack.isEmpty() && val > stack.peek()) {
                    low = stack.pop();
                }
                stack.push(val);
            }
        }
        return true;
    }
}
